core/devdoc/oop_module_requirements.md

```diff
@@ -0,0 +1,106 @@
+Module Requirements
+===================
+
+## Overview
+
+This is the documentation for an out-of-process module that communicates with
+other modules through the Azure IoT Field Gateway. Out-of-process modules need
+to implement the well defined Module Interface. Each out-of-process module will
+additionally need to attach and detach from the gateway and will either need to
+manage the worker thread manually or use the convenience methods supplied to
+start and halt the worker thread.
+
+## References
+
+[Azure IoT Field Gateway Module Interface](module.md)
+
+![High Level Design Diagrams](media/oop_module_hld.PNG)
+
+## Exposed Types
+
+```c
+typedef OOP_MODULE * OOP_MODULE_HANDLE;
+```
+
+## Exposed API
+
+### OopModule_Attach
+
+`OopModule_Attach` will provide the OopModule library with the caller's
+"Module API" and the connection key required to communicate with the Azure IoT
+Field Gateway. When `OopModule_Attach` is called the OopModule library will begin
+listening for the gateway create message. Once the create message is received
+the OopModule library will call `Module_ParseConfigurationFromJson`^, `Module_Create`
+and `Module_FreeConfiguration`^.
+*^ Assuming the optional API has been provided.*
+
+```c
+extern
+OOP_MODULE_HANDLE
+OopModule_Attach (
+    const MODULE_API * module_api_,
+    const char * const command_channel_uri_
+);
+```
+
+### OopModule_Detach
+
+`OopModule_Detach` will cease to interact with the Azure IoT Field Gateway
+and clean up all resources required to maintain the connection with the gateway.
+
+```c
+extern
+void
+OopModule_Detach (
+    OOP_MODULE_HANDLE oop_module_
+);
+```
+
+### OopModule_DoWork
+
+`OopModule_DoWork` is intended to provide the caller with fine-grain control of the
+worker thread. Therefore, `OopModule_DoWork` is a non-blocking call, which will check
+for messages from the Azure IoT Field Gateway. Each call will check for one message
+on each of the command and message channels. In other words, if multiple messages are
+queued on a single channel, only the first message of each channel will be serviced.
+If the message is intended for the out-of-process module (as opposed to the OopModule
+library itself), the OopModule library will pass it along by calling `Module_Receive`
+on the out-of-process module.
+
+```c
+extern
+void
+OopModule_DoWork (
+    OOP_MODULE_HANDLE oop_module_
+);
+```
+
+### OopModule_HaltWorkerThread
+
+`OopModule_HaltWorkerThread` will signal and join the message thread. Once
+this method has been invoked, no more messages will be received from the Azure IoT
+Field Gateway without manually calling `OopModule_DoWork`.
+
+```c
+extern
+void
+OopModule_HaltWorkerThread (
+    OOP_MODULE_HANDLE oop_module_
+);
+```
+
+### OopModule_StartWorkerThread
+
+`OopModule_StartWorkerThread` is a convenience method which gives control of
```

Comment:
I was just going to jump straight into *Olde English*, but I thought I would lose the audience.